Add publishing of dependency constraints to Gradle module metadata

jjohannes opened the pull request:
See #3551. Follow up to #3588.

bigdaz commented on subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy:
```diff
@@ -201,6 +201,67 @@ class MavenPublishJavaIntegTest extends AbstractMavenPublishIntegTest {
+                api "org.springframework:spring-core:2.5.6"
+                constraints {
+                    implementation "commons-logging:commons-logging:1.2"
+                }
```
Let's add a few more dependency constraints here, to make sure they are published correctly:
- `api` vs `implementation`
- Direct dependency with no version + dependency constraint providing version. (I guess that will fail when resolving the published POM, so might need a separate test case for now).
- Constraint for `implementation` that overrides a constraint for `api` (same module, different version)
- Constraint with strict version

jjohannes replied:
I am adding these as part of #3630 to also treat the consumption in the tests.

bigdaz commented on subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy:
```diff
@@ -201,6 +201,67 @@ class MavenPublishJavaIntegTest extends AbstractMavenPublishIntegTest {
         ]
     }
 
+    def "can publish java-library with dependency constraints"() {
```
I guess we should also test this for publishing to Ivy repository. (IvyPublishJavaIntegTest)

jjohannes replied:
Yeah I guess so... 
I am adding these as part of #3630 to also treat the consumption in the tests.

jjohannes commented on subprojects/docs/src/docs/design/gradle-module-metadata-specification.md:
```diff
@@ -101,6 +101,16 @@ Each element must be an object with the of the following values:
 
 An exclude that has a wildcard value for both `group` and `module` will exclude _all_ transitive dependencies.
 
+### `dependencyConstratins` value
```
Good catch! Fixed.

jjohannes commented on subprojects/internal-integ-testing/src/main/groovy/org/gradle/test/fixtures/GradleModuleMetadata.groovy:
```diff
@@ -161,6 +163,21 @@ class GradleModuleMetadata {
+            dependencyConstraints
+        }
+
+        Variant noMoreDependencyConstraints() {
```
Good point. I was just mirroring what we did for dependencies thinking much about it. I changed it as you suggested.

jjohannes commented on subprojects/maven/src/integTest/groovy/org/gradle/api/publish/maven/MavenPublishJavaIntegTest.groovy:
```diff
@@ -201,6 +201,67 @@ class MavenPublishJavaIntegTest extends AbstractMavenPublishIntegTest {
+                rejects()
+            }
+            noMoreDependencies()
+            noMoreDependencyConstraints()
```
Fixed.

jjohannes commented:
Thanks for the review and the very good suggestions on test coverage @bigdaz. 

I addressed the comments and merged. We now have test that show the issue with _not_ publishing constraints to maven/ivy when the version is missing. I created this follow up to address this at a later point: https://github.com/gradle/gradle/issues/3655
